File: skills/forensic-transcript-verifier/scripts/write_transcript_docx.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const path = require("path");
const {
  AlignmentType,
  BorderStyle,
  Document,
  Footer,
  Header,
  PageBreak,
  PageNumber,
  Packer,
  Paragraph,
  ShadingType,
  Table,
  TableCell,
  TableRow,
  TextRun,
  WidthType,
} = require("docx");

const FONT = "Noto Sans CJK TC";
const A4_WIDTH = 11906;
const A4_HEIGHT = 16838;
const CONTENT_WIDTH = 9026;
const BORDER = { style: BorderStyle.SINGLE, size: 1, color: "B8C2CC" };
const BORDERS = { top: BORDER, bottom: BORDER, left: BORDER, right: BORDER };

function clean(value) {
  return String(value || "").replace(/[\x00-\x08\x0B\x0C\x0E-\x1F]/g, "");
}

function textRun(text, options = {}) {
  return new TextRun({ text: clean(text), font: FONT, size: 22, ...options });
}

function cell(text, width, header = false) {
  return new TableCell({
    borders: BORDERS,
    width: { size: width, type: WidthType.DXA },
    shading: header ? { fill: "D9EAF7", type: ShadingType.CLEAR } : undefined,
    margins: { top: 90, bottom: 90, left: 120, right: 120 },
    children: [new Paragraph({ children: [textRun(text, { bold: header, size: 19 })] })],
  });
}

function unresolvedTable(rows) {
  const widths = [1500, 1400, 2500, 3626];
  const output = [new TableRow({
    tableHeader: true,
    children: [
      cell("時間", widths[0], true),
      cell("發話者", widths[1], true),
      cell("未決內容", widths[2], true),
      cell("原因／所需確認", widths[3], true),
    ],
  })];
  for (const row of rows || []) {
    output.push(new TableRow({ children: [
      cell(row.time, widths[0]),
      cell(row.speaker, widths[1]),
      cell(row.content, widths[2]),
      cell(row.reason, widths[3]),
    ] }));
  }
  return new Table({ width: { size: CONTENT_WIDTH, type: WidthType.DXA }, columnWidths: widths, rows: output });
}

async function main() {
  const taskPath = process.argv[2];
  if (!taskPath) throw new Error("usage: write_transcript_docx.js task.json");
  const data = JSON.parse(fs.readFileSync(taskPath, "utf8"));
  const outputPath = path.resolve(String(data.output_path || ""));
  if (!outputPath.toLowerCase().endsWith(".docx")) throw new Error("output_path must be .docx");
  fs.mkdirSync(path.dirname(outputPath), { recursive: true });

  const children = [
    new Paragraph({
      alignment: AlignmentType.CENTER,
      spacing: { after: 260 },
      children: [textRun(data.title || "訊問影音完整譯文（重新勘驗校正版）", { bold: true, size: 34 })],
    }),
    new Paragraph({
      spacing: { after: 140 },
      children: [textRun(data.case_info || "", { size: 20, color: "4B5563" })],
    }),
    new Paragraph({
      spacing: { after: 260 },
      children: [textRun(
        "說明：本譯文由 MAGI 依影音、時間戳 ASR、畫面序列及人工校正節文進行兩次獨立勘驗。人工校正節文於其範圍內逐字保留；不確定內容以【】明示。",
        { size: 20 }
      )],
    }),
  ];

  for (const turn of data.turns || []) {
    children.push(new Paragraph({
      keepNext: true,
      spacing: { before: 150, after: 60 },
      children: [
        textRun(`[${clean(turn.display)}] `, { bold: true, color: "1F4E79", size: 20 }),
        textRun(turn.speaker || "【發話者未定】", { bold: true, size: 20 }),
      ],
    }));
    const uncertain = /【[^】]*(?:聽辨|發話者|未定)[^】]*】/.test(clean(turn.text));
    children.push(new Paragraph({
      spacing: { after: 80, line: 360 },
      children: [textRun(`「${clean(turn.text)}」`, { color: uncertain ? "9C0006" : "111827" })],
    }));
  }

  children.push(new Paragraph({ children: [new PageBreak()] }));
  children.push(new Paragraph({
    spacing: { after: 180 },
    children: [textRun("未決事項與人工最終確認清單", { bold: true, size: 28 })],
  }));
  if ((data.unresolved || []).length) {
    children.push(unresolvedTable(data.unresolved));
  } else {
    children.push(new Paragraph({ children: [textRun("本次兩輪技術複核未留下未決項目；法院送件前仍須由承辦人確認。", { size: 20 })] }));
  }

  const doc = new Document({
    styles: {
      default: { document: { run: { font: FONT, size: 22 } } },
    },
    sections: [{
      properties: {
        page: {
          size: { width: A4_WIDTH, height: A4_HEIGHT },
          margin: { top: 1080, right: 1440, bottom: 1080, left: 1440 },
        },
      },
      headers: { default: new Header({ children: [new Paragraph({
        alignment: AlignmentType.RIGHT,
        children: [textRun(data.header || data.title || "法院影音勘驗譯文", { size: 16, color: "6B7280" })],
      })] }) },
      footers: { default: new Footer({ children: [new Paragraph({
        alignment: AlignmentType.CENTER,
        children: [textRun("— ", { size: 16 }), new TextRun({ children: [PageNumber.CURRENT], font: FONT, size: 16 }), textRun(" —", { size: 16 })],
      })] }) },
      children,
    }],
  });

  const buffer = await Packer.toBuffer(doc);
  const temporary = `${outputPath}.${process.pid}.tmp`;
  fs.writeFileSync(temporary, buffer);
  fs.renameSync(temporary, outputPath);
  process.stdout.write(JSON.stringify({ success: true, output: outputPath, turns: (data.turns || []).length }));
}

main().catch((error) => {
  process.stderr.write(String(error && error.stack ? error.stack : error));
  process.exit(1);
});

File: skills/forensic-transcript-verifier/scripts/write_transcript_docx.py
#!/usr/bin/env python3
"""Write the court transcript DOCX with the hash-bound V3 Python runtime."""

from __future__ import annotations

import json
import os
import re
import sys
from pathlib import Path
from typing import Any, Mapping

from docx import Document
from docx.enum.section import WD_SECTION
from docx.enum.table import WD_CELL_VERTICAL_ALIGNMENT, WD_TABLE_ALIGNMENT
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.oxml import OxmlElement
from docx.oxml.ns import qn
from docx.shared import Cm, Pt, RGBColor


FONT = "Noto Sans CJK TC"
CONTROL_CHARACTERS = re.compile(r"[\x00-\x08\x0B\x0C\x0E-\x1F]")
UNCERTAIN = re.compile(r"【[^】]*(?:聽辨|發話者|未定)[^】]*】")


def clean(value: Any) -> str:
    return CONTROL_CHARACTERS.sub("", str(value or ""))


def set_run_font(run: Any, *, size: float, bold: bool = False, color: str | None = None) -> None:
    run.font.name = FONT
    run.font.size = Pt(size)
    run.bold = bold
    run._element.get_or_add_rPr().rFonts.set(qn("w:eastAsia"), FONT)
    if color:
        run.font.color.rgb = RGBColor.from_string(color)


def add_run(
    paragraph: Any,
    text: Any,
    *,
    size: float = 11,
    bold: bool = False,
    color: str | None = None,
) -> Any:
    run = paragraph.add_run(clean(text))
    set_run_font(run, size=size, bold=bold, color=color)
    return run


def set_cell_shading(cell: Any, fill: str) -> None:
    properties = cell._tc.get_or_add_tcPr()
    shading = properties.find(qn("w:shd"))
    if shading is None:
        shading = OxmlElement("w:shd")
        properties.append(shading)
    shading.set(qn("w:fill"), fill)


def set_cell_text(cell: Any, value: Any, *, header: bool = False) -> None:
    cell.text = ""
    cell.vertical_alignment = WD_CELL_VERTICAL_ALIGNMENT.CENTER
    paragraph = cell.paragraphs[0]
    paragraph.paragraph_format.space_after = Pt(0)
    add_run(paragraph, value, size=9.5, bold=header)
    if header:
        set_cell_shading(cell, "D9EAF7")


def add_page_number(paragraph: Any) -> None:
    add_run(paragraph, "— ", size=8, color="6B7280")
    begin = OxmlElement("w:fldChar")
    begin.set(qn("w:fldCharType"), "begin")
    instruction = OxmlElement("w:instrText")
    instruction.set(qn("xml:space"), "preserve")
    instruction.text = " PAGE "
    separate = OxmlElement("w:fldChar")
    separate.set(qn("w:fldCharType"), "separate")
    value = OxmlElement("w:t")
    value.text = "1"
    end = OxmlElement("w:fldChar")
    end.set(qn("w:fldCharType"), "end")
    run = paragraph.add_run()
    set_run_font(run, size=8, color="6B7280")
    run._r.extend((begin, instruction, separate, value, end))
    add_run(paragraph, " —", size=8, color="6B7280")


def write_document(task: Mapping[str, Any]) -> Path:
    output = Path(clean(task.get("output_path"))).expanduser().resolve()
    if output.suffix.lower() != ".docx":
        raise ValueError("output_path must be .docx")
    output.parent.mkdir(parents=True, exist_ok=True)

    document = Document()
    section = document.sections[0]
    section.start_type = WD_SECTION.NEW_PAGE
    section.page_width = Cm(21)
    section.page_height = Cm(29.7)
    section.top_margin = Cm(1.9)
    section.right_margin = Cm(2.54)
    section.bottom_margin = Cm(1.9)
    section.left_margin = Cm(2.54)

    normal = document.styles["Normal"]
    normal.font.name = FONT
    normal.font.size = Pt(11)
    normal._element.get_or_add_rPr().rFonts.set(qn("w:eastAsia"), FONT)
    zoom = document.settings.element.find(qn("w:zoom"))
    if zoom is not None:
        # python-docx emits bestFit without the percent attribute required by
        # the OOXML validator used by MAGI's court-grade DOCX gate.
        zoom.set(qn("w:percent"), "100")

    title = clean(task.get("title") or "訊問影音完整譯文（重新勘驗校正版）")
    header_text = clean(task.get("header") or title or "法院影音勘驗譯文")
    header = section.header.paragraphs[0]
    header.alignment = WD_ALIGN_PARAGRAPH.RIGHT
    add_run(header, header_text, size=8, color="6B7280")
    footer = section.footer.paragraphs[0]
    footer.alignment = WD_ALIGN_PARAGRAPH.CENTER
    add_page_number(footer)

    paragraph = document.add_paragraph()
    paragraph.alignment = WD_ALIGN_PARAGRAPH.CENTER
    paragraph.paragraph_format.space_after = Pt(13)
    add_run(paragraph, title, size=17, bold=True)

    paragraph = document.add_paragraph()
    paragraph.paragraph_format.space_after = Pt(7)
    add_run(paragraph, task.get("case_info"), size=10, color="4B5563")

    paragraph = document.add_paragraph()
    paragraph.paragraph_format.space_after = Pt(13)
    add_run(
        paragraph,
        "說明：本譯文由 MAGI 依影音、時間戳 ASR、畫面序列及人工校正節文進行兩次獨立勘驗。"
        "人工校正節文於其範圍內逐字保留；不確定內容以【】明示。",
        size=10,
    )

    turns = task.get("turns") if isinstance(task.get("turns"), list) else []
    for raw in turns:
        turn = raw if isinstance(raw, Mapping) else {}
        heading = document.add_paragraph()
        heading.paragraph_format.keep_with_next = True
        heading.paragraph_format.space_before = Pt(7.5)
        heading.paragraph_format.space_after = Pt(3)
        add_run(heading, f"[{clean(turn.get('display'))}] ", size=10, bold=True, color="1F4E79")
        add_run(heading, turn.get("speaker") or "【發話者未定】", size=10, bold=True)

        body = document.add_paragraph()
        body.paragraph_format.space_after = Pt(4)
        body.paragraph_format.line_spacing = 1.5
        text = clean(turn.get("text"))
        add_run(body, f"「{text}」", size=11, color="9C0006" if UNCERTAIN.search(text) else "111827")

    document.add_page_break()
    paragraph = document.add_paragraph()
    paragraph.paragraph_format.space_after = Pt(9)
    add_run(paragraph, "未決事項與人工最終確認清單", size=14, bold=True)

    unresolved = task.get("unresolved") if isinstance(task.get("unresolved"), list) else []
    if unresolved:
        table = document.add_table(rows=1, cols=4)
        table.alignment = WD_TABLE_ALIGNMENT.CENTER
        table.style = "Table Grid"
        widths = (Cm(2.5), Cm(2.3), Cm(4.2), Cm(6.0))
        labels = ("時間", "發話者", "未決內容", "原因／所需確認")
        for index, label in enumerate(labels):
            table.columns[index].width = widths[index]
            set_cell_text(table.rows[0].cells[index], label, header=True)
        for raw in unresolved:
            row = raw if isinstance(raw, Mapping) else {}
            cells = table.add_row().cells
            values = (row.get("time"), row.get("speaker"), row.get("content"), row.get("reason"))
            for index, value in enumerate(values):
                cells[index].width = widths[index]
                set_cell_text(cells[index], value)
    else:
        paragraph = document.add_paragraph()
        add_run(paragraph, "本次兩輪技術複核未留下未決項目；法院送件前仍須由承辦人確認。", size=10)

    temporary = output.with_name(f"{output.name}.{os.getpid()}.tmp")
    try:
        document.save(temporary)
        os.replace(temporary, output)
    finally:
        temporary.unlink(missing_ok=True)
    return output


def main() -> int:
    if len(sys.argv) != 2:
        raise ValueError("usage: write_transcript_docx.py task.json")
    task_path = Path(sys.argv[1]).expanduser().resolve(strict=True)
    payload = json.loads(task_path.read_text(encoding="utf-8"))
    if not isinstance(payload, dict):
        raise ValueError("task JSON must be an object")
    output = write_document(payload)
    print(
        json.dumps(
            {"success": True, "output": str(output), "turns": len(payload.get("turns") or [])},
            ensure_ascii=False,
            sort_keys=True,
        )
    )
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:
        print(str(exc), file=sys.stderr)
        raise SystemExit(1)
